src/Storages/MergeTree/MergeTreeIndexIVFFlat.cpp

```diff
@@ -0,0 +1,260 @@
+{
+    detail::ReadBufferFaissWrapper istr_wrapped(istr);
+    FaissBaseIndex* index = faiss::read_index(&istr_wrapped);
+    index_base.reset(index);
+}
+
+bool MergeTreeIndexGranuleIVFFlat::empty() const
+{
+    // return false;
+    return index_base->ntotal == 0;
+}
+
+
+MergeTreeIndexAggregatorIVFFlat::MergeTreeIndexAggregatorIVFFlat(const String & index_name_,
+                                                                const Block & index_sample_block_)
+    : index_name(index_name_)
+    , index_sample_block(index_sample_block_)
+{}
+
+bool MergeTreeIndexAggregatorIVFFlat::empty() const
+{
+    // return true;
+    return values.empty();
+}
+
+MergeTreeIndexGranulePtr MergeTreeIndexAggregatorIVFFlat::getGranuleAndReset()
+{
+    // Build index
+    size_t dim = 3;
+    size_t nlist = 5;
+
+    // IndexIFVFlat will take care of memory managment
+    MergeTreeIndexGranuleIVFFlat::FaissBaseIndex* coarse_quantizer = new faiss::IndexFlat(dim, faiss::METRIC_L2);
+    // faiss::IndexFlatL2 coarse_quantizer(dim);
+    auto index = std::make_shared<faiss::IndexIVFFlat>(coarse_quantizer, dim, nlist);
+    index->own_fields = true;
+    index->verbose = true;
+
+    // Take care about Float32 == float
+
+    for (size_t i = 0; i < values.size(); i += 200) {
+        LOG_DEBUG(&Poco::Logger::get("IVFFlat"), "i = {}, val = {}", i, values[i]);
+    }
+
+    auto num_elements = values.size() / dim;
+    LOG_DEBUG(&Poco::Logger::get("IVFFlat"), "Val size: {}", num_elements);
+    index->train(num_elements, values.data());
+    LOG_DEBUG(&Poco::Logger::get("IVFFlat"), "Train completed");
+    index->add(num_elements, values.data());
+    LOG_DEBUG(&Poco::Logger::get("IVFFlat"), "Add completed");
+
+    values.clear();
+
+    return std::make_shared<MergeTreeIndexGranuleIVFFlat>(index_name, index_sample_block, index);
+}
+
+void MergeTreeIndexAggregatorIVFFlat::update(const Block & block, size_t * pos, size_t limit)
+{
+    if (*pos >= block.rows())
+        throw Exception(
+                "The provided position is not less than the number of block rows. Position: "
+                + toString(*pos) + ", Block rows: " + toString(block.rows()) + ".", ErrorCodes::LOGICAL_ERROR);
+
+    size_t rows_read = std::min(limit, block.rows() - *pos);
+
+    // This index can be used only for one column
+    assert(index_sample_block.columns() == 1);
+
+    auto index_column_name = index_sample_block.getByPosition(0).name;
+    const auto & column = block.getByName(index_column_name).column->cut(*pos, rows_read);
+
+    for (size_t i = 0; i < rows_read; ++i) {
+        Field field;
+        column->get(i, field);
+        
+        auto field_array = field.safeGet<Tuple>();
+
+        // Store vectors in the flatten arrays
+        for (const auto& value : field_array) {
+            auto num = value.safeGet<Float32>();
+            values.push_back(num);
+        }
+    }
```

Comment:
It is better to do `typeid_cast`  to `const ColumnTuple *`, then get vector `Columns` and apply `typeid_cast` for each element to `const ColumnFolat64 *`, and then call `getData`, which will return PodArray<Float64>